test/cql-pytest/test_in_operator.py

```diff
@@ -0,0 +1,80 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 ScyllaDB
+#
+# This file is part of Scylla.
+#
+# Scylla is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Scylla is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+
+import pytest
+from cassandra import InvalidRequest
+from cassandra.cluster import Session, NoHostAvailable
+from cassandra.protocol import SyntaxException
+
+from util import unique_name
+
+
+@pytest.fixture(scope="module")
+def table_name(request, cql: Session, test_keyspace: str) -> str:
+    table = f"{test_keyspace}.{unique_name()}"
+
+    cql.execute(f"CREATE TABLE {table} (p int PRIMARY KEY, a ascii, b boolean, d double,"
+                f"mai map<ascii, int>, tup frozen<tuple<text, int>>, l list<text>)")
+    request.addfinalizer(lambda: cql.execute("DROP TABLE " + table))
+    return table
+
+
+@pytest.mark.parametrize("column,column_type", [
+    ("mai", "map<ascii, int>"),
+    ("l", "list<text>"),
+])
+def test_in_query_with_invalid_column_type(cql: Session, table_name: str, column: str, column_type: str):
+    expected_error = 'Error from server: code=2200 [Invalid query] message="' \
+                     f'Collection column \'{column}\' ({column_type}) cannot be restricted by a \'IN\' relation"'
+    with pytest.raises(InvalidRequest) as error:
+        cql.prepare(f"SELECT * FROM {table_name} where {column} IN (?, ?) ALLOW FILTERING;")
+    assert error.value.args[0] == expected_error
+
+
+def test_in_query_without_allow_filtering(cql: Session, table_name: str):
+    expected_error = 'Error from server: code=2200 [Invalid query] message="' \
+                     f'Cannot execute this query as it might involve data filtering and thus may have unpredictable' \
+                     f' performance. If you want to execute this query despite the performance unpredictability,' \
+                     f' use ALLOW FILTERING"'
+    with pytest.raises(InvalidRequest) as error:
+        cql.prepare(f"SELECT * FROM {table_name} where p IN (?, ?) AND a IN (?, ?);")
+    assert error.value.args[0] == expected_error
+
+
+def test_in_query_with_valid_product_limit(cql: Session, table_name: str):
+    limit = 100
+    query = f"SELECT * FROM {table_name} WHERE p IN {tuple(idx for idx in range(limit))} ALLOW FILTERING;"
+    cql.execute(query)
+
+
+def test_in_query_with_invalid_product_limit(cql: Session, table_name: str):
+    limit = 101
```

Comment:
I'm not happy about such a test. Not only will it not fail as you expect on Cassandra (so should be marked scylla_only, if we keep it), I also don't see any reason to enshrine this limit in a test. Why do we need a test to fail if a year from now somebody will increase the limit to 200?
Another option is to use a very large number here-  e.g., 1,000,000 - and this should fail even if we increase the limit. The problem with this option is that it will send a huge request and can be very slow. I also don't know what it will do on Cassandra (we'll need to check).

Reply:
Done